### DISCUSSION TITLE
[Spindle A/B/Z Rotary Encoder] - Minimum PPR requirement?

### DISCUSSION THREAD

**Discussion by GTCLive** [Peter-van-Tol/LiteX-CNC#61] [Q&A] [answered]
I'm having difficulty reading my spindle RPM with my old Rotary A/B/Z 100ppr encoder (which works fine on a parport config). I'm getting the feeling that 100PPR ain't sufficient for Litex-CNC. My FPGA is 40Mhz and the linuxcnc servo thread is 1000000. What sort of encoder PPR would be best to read 0~5000 RPM?

**Comment by Peter-van-Tol** [comment, marked as answer]
To be honest: anything with an index pulse would work for a spindle. When the speed is constant, the index pulse can be used to start the movements.

LitexCNC does not have any requirements on the resolution of the encoder. Can you describe why you think that the 100PPR encoder is not sufficient? Maybe something is wrong and needs improvement on the LitexCNC side.

If you want to go fancy and really make 'digital gears', then I would recommend a 2500 PPR encoder. In that case it is possible to create non-round objects with your lathe.

**Reply by GTCLive** [reply]
Are you sure 100ppr is sufficient? For a starter I have no hal config reference, experience neither the knowledge to rig a hal config together to test this out and so the problem could well be anything at this stage. Secondly LitexCNC seems fine, I can see the 100 counts if I rotate the spindle one rotation by hand, issue appears to be the velocity/rpm calcs - I'm only getting fluctuation in the low positive and negative decimals. I was suspecting 'not enough of 400 pulses' to compute...

Below code is what I'm using to test. PWM and Spindle works fine. Just can't get any RPM readout. 

```
###Spindle test
net spindle-speed-cmd spindle.0.speed-out => [LITEXCNC](NAME).pwm.00.value
net spindle-on spindle.0.on => [LITEXCNC](NAME).pwm.00.enable
setp [LITEXCNC](NAME).pwm.00.pwm_freq 	2000      ### 2Khz pulse
setp [LITEXCNC](NAME).pwm.00.min_dc 	0.09      ### Duty cycle min. 
setp [LITEXCNC](NAME).pwm.00.max_dc  	0.99      ### Duty cycle man. 
setp [LITEXCNC](NAME).pwm.00.scale		2000.0      ### Max RPM. 
net spindle-ccw spindle.0.reverse => [LITEXCNC](NAME).gpio.27.out

###Encoder01
setp [LITEXCNC](NAME).encoder.01.position-scale 400.0      ### 100ppr rotary encoder
setp [LITEXCNC](NAME).encoder.01.x4-mode 1
net spindle-revs             <=   [LITEXCNC](NAME).encoder.01.position
net spindle-vel-fb-rps       <=   [LITEXCNC](NAME).encoder.01.velocity
net spindle-vel-fb-rpm       <=   [LITEXCNC](NAME).encoder.01.velocity-rpm
net spindle-index-enable     <=>  [LITEXCNC](NAME).encoder.01.index-enable
```

**Reply by GTCLive** [reply]
Peter, hold-on. I'm went to the shop to oscope'd the inputs of the encoder01 to rule that part out and found an erratic Z index pulse/signal. Machine ground's bad or something (wild guess). I'll update in a few hours...

**Reply by GTCLive** [reply]
Yeah. Works now. Still got a nasty 80khz -0.2(!!)~0.2 volt stray in the system but that's irrelevant to LitexCNC. Those cheap SMPS are a pita... might have to swap that for a linear power supply. Endless...

Anyway. Here's a functional HAL and other bits that seems to work with a 100PPR. Not using PID but lincurve here. My experience with PID for a closed loop spindle is, it sucks. Always over/under shoots and takes forever to settle. Lincurve is instant/better (I think). 

### HAL
```
loadrt [KINS]KINEMATICS
loadrt [EMCMOT]EMCMOT servo_period_nsec=[EMCMOT]SERVO_PERIOD num_joints=[KINS]JOINTS
loadrt litexcnc connections="eth:192.168.10.101"

loadrt abs count=1
loadrt scale count=1
loadrt lowpass count=1
loadrt near

### Setup the watchdog (assuming servo-thread period of 1000000 ns (1 kHz))
setp [LITEXCNC](NAME).watchdog.timeout_ns 5000000

addf [LITEXCNC](NAME).read		servo-thread
addf motion-command-handler		servo-thread
addf motion-controller			servo-thread

addf abs.0 				servo-thread
addf scale.0 				servo-thread
addf lowpass.0 				servo-thread
addf near.0 				servo-thread
# ...
addf [LITEXCNC](NAME).write   servo-thread


### SPINDLE
setp [LITEXCNC](NAME).pwm.00.pwm_freq		2000
setp [LITEXCNC](NAME).pwm.00.scale		1562.5
setp [LITEXCNC](NAME).pwm.00.dither_pwm		true
net spindle-on <= spindle.0.on => [LITEXCNC](NAME).pwm.00.enable
net spindle-cmd-rpm => [LITEXCNC](NAME).pwm.00.value			
net spindle-cmd-rps <= spindle.0.speed-out-rps
net spindle-cmd-rps-abs <= spindle.0.speed-out-rps-abs
net spindle-at-speed => spindle.0.at-speed
net spindle-ccw spindle.0.reverse => [LITEXCNC](NAME).gpio.27.out

### Rotary Encoder (100ppr)
setp [LITEXCNC](NAME).encoder.01.position-scale 400.0
setp [LITEXCNC](NAME).encoder.01.x4-mode 1
net spindle-position [LITEXCNC](NAME).encoder.01.position => spindle.0.revs
net spindle-velocity-feedback-rps [LITEXCNC](NAME).encoder.01.velocity => spindle.0.speed-in
net spindle-index-enable <=> spindle.0.index-enable	[LITEXCNC](NAME).encoder.01.index-enable
```
### pyvcp_option.hal
```
###Spindle at speed indicator
setp scale.0.gain 60
setp lowpass.0.gain 0.010000
net spindle-velocity-feedback-rps	=> lowpass.0.in
net spindle-fb-filtered-rps	lowpass.0.out => abs.0.in
net spindle-fb-filtered-abs-rps	abs.0.out => scale.0.in
net spindle-fb-filtered-abs-rpm	scale.0.out => pyvcp.spindle-speed

net spindle-cmd-rps	=>  near.0.in1
net spindle-velocity-feedback-rps	=>  near.0.in2
net spindle-at-speed	<=  near.0.out
setp near.0.scale	1.500000
net spindle-at-speed	=> pyvcp.spindle-at-speed-led
```
### custom.hal 
```
###Spindle Lincurve
loadrt lincurve personality=15
addf lincurve.0 servo-thread
setp lincurve.0.x-val-00 100
setp lincurve.0.y-val-00 140
setp lincurve.0.x-val-01 200
setp lincurve.0.y-val-01 235
setp lincurve.0.x-val-02 300
setp lincurve.0.y-val-02 335
setp lincurve.0.x-val-03 400
setp lincurve.0.y-val-03 428
setp lincurve.0.x-val-04 500
setp lincurve.0.y-val-04 510
setp lincurve.0.x-val-05 600
setp lincurve.0.y-val-05 590
setp lincurve.0.x-val-06 700
setp lincurve.0.y-val-06 670
setp lincurve.0.x-val-07 800
setp lincurve.0.y-val-07 743
setp lincurve.0.x-val-08 900
setp lincurve.0.y-val-08 810
setp lincurve.0.x-val-09 1000
setp lincurve.0.y-val-09 875
setp lincurve.0.x-val-10 1200
setp lincurve.0.y-val-10 990
setp lincurve.0.x-val-11 1400
setp lincurve.0.y-val-11 1098
setp lincurve.0.x-val-12 1600
setp lincurve.0.y-val-12 1190
setp lincurve.0.x-val-13 1800
setp lincurve.0.y-val-13 1275
setp lincurve.0.x-val-14 2000
setp lincurve.0.y-val-14 1350
net spindle-corr <= spindle.0.speed-out-abs lincurve.0.in
net spindle-cmd-rpm <= lincurve.0.out
```
### INI
```
[SPINDLE_0]
MAX_FORWARD_VELOCITY = 2000
MIN_FORWARD_VELOCITY = 125
MAX_REVERSE_VELOCITY = 2000
MIN_REVERSE_VELOCITY = 125
INCREMENT = 100
```

**Comment by GTCLive** [comment]
### HAL Example of a PWM Closed Loop Spindle with LitexCNC 

For Reference only. I was toying with the encoder and spindle yesterday and had this working pretty well using a PWM output and a 100ppr encoder. Fully Functional with Sxxx M3 however (M4-abs) remains to be done along with filtering, perhaps. If anybody get to finish it up please update/share. 


### HAL
```
loadrt [KINS]KINEMATICS
loadrt [EMCMOT]EMCMOT servo_period_nsec=[EMCMOT]SERVO_PERIOD num_joints=[KINS]JOINTS
loadrt litexcnc connections="eth:192.168.10.101"

loadrt abs count=1
loadrt scale count=1
loadrt near
loadrt pid num_chan=1
loadrt sum2 count=1

setp [LITEXCNC](NAME).watchdog.timeout_ns 5000000

addf [LITEXCNC](NAME).read		servo-thread
addf motion-command-handler		servo-thread
addf motion-controller			servo-thread

addf abs.0 				servo-thread
addf scale.0 				servo-thread
addf near.0 				servo-thread
addf pid.0.do-pid-calcs 		servo-thread
addf sum2.0 				servo-thread
addf [LITEXCNC](NAME).write   servo-thread


########################################################################
### PWM SPINDLE (CLOSED LOOP with 100PPR rotary/incremental encoder)
########################################################################
setp pid.0.Pgain 	0
setp pid.0.Igain 	2
setp pid.0.Dgain 	0
setp pid.0.FF0 		0
setp pid.0.FF1 		0
setp pid.0.bias      	0
setp pid.0.deadband  	5
setp pid.0.maxerrorI 	100
setp pid.0.maxerror 	0
setp pid.0.maxoutput	2000
setp pid.0.error-previous-target	true

net spindle-cmd-rpm => [LITEXCNC](NAME).pwm.00.value
net spindle-on <= spindle.0.on => [LITEXCNC](NAME).pwm.00.enable => pid.0.enable

setp [LITEXCNC](NAME).pwm.00.pwm_freq 		2000
setp [LITEXCNC](NAME).pwm.00.scale 		2000
setp [LITEXCNC](NAME).pwm.00.min_dc 		0.09
setp [LITEXCNC](NAME).pwm.00.max_dc  		0.99
setp [LITEXCNC](NAME).pwm.00.dither_pwm 	true  #no idea what that does. 
	
net spindle-cmd-rpm     <= sum2.0.out
net spindle-cmd-rpm-abs <= spindle.0.speed-out-abs
net spindle-cmd-rps     <= spindle.0.speed-out-rps
net spindle-cmd-rps-abs <= spindle.0.speed-out-rps-abs
net spindle-at-speed    => spindle.0.at-speed

setp [LITEXCNC](NAME).encoder.01.position-scale 400.0
setp [LITEXCNC](NAME).encoder.01.x4-mode 1
setp scale.0.gain 60
net spindle-position [LITEXCNC](NAME).encoder.01.position => spindle.0.revs
net spindle-velocity-feedback-rps [LITEXCNC](NAME).encoder.01.velocity => scale.0.in
net spindle-index-enable [LITEXCNC](NAME).encoder.01.index-enable <=> spindle.0.index-enable

net pid_feedback pid.0.feedback <= spindle.0.speed-in <= scale.0.out
net pid_command pid.0.command <= spindle.0.speed-out => sum2.0.in0
net pid_output pid.0.output => sum2.0.in1

setp near.0.scale	1.1
net spindle-cmd-rps	=>  near.0.in1
net spindle-velocity-feedback-rps	=>  near.0.in2
net spindle-at-speed	<=  near.0.out
```
### pyvcp_option.hal
```
net pid_feedback	pyvcp.spindle-speed
net spindle-at-speed	=> pyvcp.spindle-at-speed-led
```

**Reply by EhsanRasouli1371** [reply]
Hi, I have just joined the Linux CNC forum and to start with, I am planning to control a 6-axis robot with Litex-CNC and 5A-75E8.2 board, but I still don't know how to configure the Linux CNC files.
Is it possible for you to share the configuration files of your CNC machine with me?
Thanks

**Reply by GTCLive** [reply]
Hi Ehsan. My configs were based on Peter's modules documentation examples and quite machine specific. Although fairly simplified and left to only what is needed i.e. two motors with fb, two switches, a mpg and a probe signal. Nothing else. Not even using the software estop, all that is mechanical in-line with the main. 

That said I'm not entirely sure how this would be compatible with a complex 6-axis, perhaps be more misleading than anything else. I recommend you start enabling the motors using the example in the doc for a starter, then enable the switches, and slowly enable the rest as needed. Note that some HAL and Python learning and experimenting might be required (of which I knew nothing about when I started this project) so don't rush anything and take it slow.

**Reply by EhsanRasouli1371** [reply]
Actually, I want to create my own robot configuration by having an example and editing it.
I think I will get results faster this way

**Reply by GTCLive** [reply]
I'm not convinced it will go faster if you don't learn and experiment before. Things normally go backward in these circumstances rather. To each his/her own way I guess. 

Peter drafted some documentation which include pointers to a successful machine config. For examples you can use several configs from the Linuxcnc installation examples (MESA has many of them). That's what I did and where I've learn from.

**Reply by Peter-van-Tol** [reply]
I can try to make a configuration for a 6 axis machine with:
- 6 stepgen
- 6 encoders
- GPIO on remaining pins.

However, this is depending on your requirements. Please create a separate subject for this, to keep the other topics clean for future users.

Where I cannot help though, is the HAL configuration in LinuxCNC. My experience is limited to lathes and routers. So you have to experiment yourself in the HAL or ask directions from forum.linuxcnc.org.

**Reply by EhsanRasouli1371** [reply]
Thank you my dear friend
It is a very good idea
I made a 6-axis robot with a Mach 3 controller and I plan to use your project instead of the Mach 3.
And I will do my best to make it happen with your help

**Reply by EhsanRasouli1371** [reply]
We can create a sample config similar to the one created for Misa or pci boards

**Reply by EhsanRasouli1371** [reply]
I can do this by editing the sample configuration in sim/axis/vismach/puma560
